File: my-folder/0068-text-justification/solution.java
class Solution {
    public  List<String> fullJustify(String[] words, int maxWidth) {
    List<String> result = new ArrayList<>();

    int i = 0;
    while (i < words.length) {
      int j = i + 1;
      int curLen = words[i].length();
      while (j < words.length && (curLen + words[j].length() + (j - i - 1) < maxWidth)) {
        curLen += words[j].length();
        j++;
      }
      int totalSpace = maxWidth - curLen;
      // one word or last line, left justify
      if (j - i == 1 || j == words.length) {
        result.add(leftJustify(words, i, j, totalSpace));
      } else { // otherwise, middle justify
        result.add(middleJustify(words, i, j, totalSpace));
      }
      i = j;
    }

    return result;
  }

  private  String middleJustify(String[] words, int i, int j, int totalSpace) {
    int sections = j - i - 1;
    // space in between
    int spaces = totalSpace / sections;
    // left over
    int extraSpaces = totalSpace % sections;

    StringBuilder sb = new StringBuilder(words[i]);
    for (int k = i + 1; k < j; k++) {
      int spacesToApply = spaces + (extraSpaces > 0 ? 1 : 0);
      extraSpaces--;
      sb.append(fillSpaces(spacesToApply)).append(words[k]);
    }

    return sb.toString();
  }
    
  private  String leftJustify(String[] words, int i, int j, int totalSpace) {
    StringBuilder sb = new StringBuilder(words[i]);
    for (int k = i + 1; k < j; k++) {
      sb.append(" ").append(words[k]);
    }
    int rightSpace = totalSpace - (j - i - 1);
    sb.append(fillSpaces(rightSpace));
    return sb.toString();
  }

  private  String fillSpaces(int n) {
    if (n <= 0) {
      return "";
    }
    char[] chars = new char[n];
    Arrays.fill(chars, ' ');
    return String.valueOf(chars);
  }
}

File: my-folder/0068-text-justification/solution.py
class Solution(object):
    def fullJustify(self, words, maxWidth):
        """
        :type words: List[str]
        :type maxWidth: int
        :rtype: List[str]
        """
        res = []
        now_len = 0
        for word in words:
            if now_len == 0:
                now_len = len(word)
                temp_res = [word]
                temp_word_count = 1       
            elif now_len+len(word)+1<maxWidth:
                temp_res+=[word]
                temp_word_count+=1
                now_len += (len(word)+1)
            elif now_len+len(word)+1==maxWidth:
                temp_res+=[word]
                temp_str = temp_res[0]               
                for i in range(1, temp_word_count+1):
                    temp_str+=" "+temp_res[i]  
                res+=[temp_str]
                temp_word_count=0
                now_len = 0
                temp_res = []
            else:
                temp_str = temp_res[0]
                if temp_word_count==1:
                    temp_str += (maxWidth-now_len)*" "
                else:
                    space_len = (maxWidth-now_len)//(temp_word_count-1)
                    space_left = (maxWidth-now_len)%(temp_word_count-1)
                    for i in range(1, temp_word_count):
                        if space_left>0:
                            temp_str+=(2+space_len)*" "+temp_res[i]
                            space_left-=1
                        else:
                            temp_str+=(space_len+1)*" "+temp_res[i]
                res+=[temp_str]
                now_len = len(word)
                temp_res = [word]
                temp_word_count = 1
        if now_len>0:
            temp_str = " ".join(temp_res)
            temp_str+=(maxWidth-now_len)*" "
            # temp_str = temp_res[0]
            # if temp_word_count==1:
            #     temp_str += (maxWidth-now_len)*" "
            # else:
            #     space_len = (maxWidth-now_len)//(temp_word_count-1)
            #     space_left = (maxWidth-now_len)%(temp_word_count-1)
            #     for i in range(1, temp_word_count):
            #         if space_left>0:
            #             temp_str+=(2+space_len)*" "+temp_res[i]
            #             space_left-=1
            #         else:
            #             temp_str+=(space_len+1)*" "+temp_res[i]
            res+=[temp_str]
            
        return res
                            
                
                    
        
